src/circuit_proof/verifier.rs

```diff
@@ -241,83 +235,77 @@ impl<'a, 'b> VerifierCS<'a, 'b> {
 
         let w = self.transcript.challenge_scalar(b"w");
 
-        let r = Scalar::random(&mut rng);
-        let xx = x * x;
-        let y_inv = y.invert();
+        let (wL, wR, wO, wV, wc) = self.flattened_constraints(&z);
 
-        // Calculate points that represent the matrices
-        let H_prime: Vec<RistrettoPoint> = gens
-            .H(n)
-            .zip(util::exp_iter(y_inv))
-            .map(|(H_i, exp_y_inv)| H_i * exp_y_inv)
-            .collect();
+        // Get IPP variables
+        let (x_sq, x_inv_sq, s) = proof.ipp_proof.verification_scalars(self.transcript);
 
-        // W_L_point = <h * y^-n , z * z^Q * W_L>, line 81
-        let W_L_point = RistrettoPoint::vartime_multiscalar_mul(&z_zQ_WL, &H_prime);
+        let a = proof.ipp_proof.a;
+        let b = proof.ipp_proof.b;
 
-        // W_R_point = <g , y^-n * z * z^Q * W_R>, line 82
-        let y_n_z_zQ_WR = z_zQ_WR
+        let y_inv = y.invert();
+        let y_inv_vec = util::exp_iter(y_inv).take(n).collect::<Vec<Scalar>>();
+        let yneg_wR = wR
             .iter()
-            .zip(util::exp_iter(y.invert()))
-            .map(|(W_R_right_i, exp_y_inv)| W_R_right_i * exp_y_inv)
+            .zip(y_inv_vec.iter())
+            .map(|(wRi, exp_y_inv)| wRi * exp_y_inv)
             .collect::<Vec<Scalar>>();
-        let W_R_point = RistrettoPoint::vartime_multiscalar_mul(&y_n_z_zQ_WR, gens.G(n));
 
-        // W_O_point = <h * y^-n , z * z^Q * W_O>, line 83
-        let W_O_point = RistrettoPoint::vartime_multiscalar_mul(&z_zQ_WO, &H_prime);
-
-        // Get IPP variables
-        let (x_sq, x_inv_sq, s) = proof.ipp_proof.verification_scalars(self.transcript);
-        let s_inv = s.iter().rev().take(n);
-        let a = proof.ipp_proof.a;
-        let b = proof.ipp_proof.b;
+        let delta = inner_product(&yneg_wR, &wL);
 
         // define parameters for P check
-        let g = s.iter().take(n).map(|s_i| -a * s_i);
-        let h = s_inv
-            .zip(util::exp_iter(y.invert()))
-            .map(|(s_i_inv, exp_y_inv)| -exp_y_inv * b * s_i_inv - Scalar::one());
+        let g = yneg_wR
+            .iter()
+            .zip(s.iter().take(n))
+            .map(|(yneg_wRi, s_i)| x * yneg_wRi - a * s_i);
+
+        let h = y_inv_vec
+            .iter()
+            .zip(s.iter().rev().take(n))
+            .zip(wL.iter())
+            .zip(wO.iter())
+            .map(|(((y_inv_i, s_i_inv), wLi), wOi)| {
+                y_inv_i * (x * wLi + wOi - b * s_i_inv) - Scalar::one()
+            });
+
+        // Create a `TranscriptRng` from the transcript
+        use rand::thread_rng;
+        let mut rng = self.transcript.build_rng().finalize(&mut thread_rng());
+        let r = Scalar::random(&mut rng);
```

Comment:
the writeup uses `c` as the random challenge variable for combining the equations, so can we change this to match the writeup? (or the other way around)

Reply:
actually I think using `r` is less confusing than using `c`, since we also have `c` representing the constants... I'll leave a note for that on the writeup